### PR TITLE
[MDS-4986]  - change keycloak config for minespace and core web

### PR DESCRIPTION
**Pull request by wilson496**
## Objective 

[MDS-4838](https://bcmines.atlassian.net/browse/MDS-4838)
[MDS-4986](https://bcmines.atlassian.net/browse/MDS-4986)

- Update to use Gold SSO 
- Implement Dual IDP support on core-api with jwtManager selector
- Update to copy core of `flask-jwt-oidc` library into mds code base and add support for multi jwt client support.

### CODE REVIEW

File: bin/compare.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+diff -r -C 5 services/core-web/common services/minespace-web/common
+
+if [ $? != 0 ]
+then 
+echo "'/services/core-web/common' does not match '/services/minespace-web/common'"
```
Comment (matbusby-fw) [inline bin/compare.sh:7]: I look forward to the day this is gone.

File: bin/compare.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
Comment (vyasworks) [inline bin/compare.sh:1]: Added this coz there were some errors but seeing them in GHA was hard, and no way to easily recreate them locally

File: services/core-api/app/__init__.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from flask_cors import CORS
 from flask_restplus import Resource, apidoc
 from sqlalchemy.exc import SQLAlchemyError
-from flask_jwt_oidc.exceptions import AuthError
+from app.flask_jwt_oidc_local.exceptions import AuthError
```
Comment (vyasworks) [inline services/core-api/app/__init__.py:9]: we are switching to the local version of this library, with enhancements. 

This is the main change to support multiple integration sso clients. 

https://github.com/thorwolpert/flask-jwt-oidc library has been built to work with only one single JWT validator and client. 

`app.flas_jwt_oidc_local` is a copy of the main jwtManager() class from the above library, and it is extended to be initiated with JWT env vars. More expln. in that section.

File: services/core-api/app/__init__.py [hunk 4]
```diff
@@ -102,7 +102,12 @@ def register_extensions(app):
     root_api_namespace.init_app(app)
 
     try:
-        jwt.init_app(app)
+
+        if test_config is None:
+            jwtv2.init_app(app)
```
Comment (vyasworks) [inline services/core-api/app/__init__.py:107]: v1 is existing keycloack 
v2 is Gold SSO. 

we will have more added here once we get integration clients in gold sso 

like: 

`jwtBCMI` `jwtDataBC` `jwtGenTx` etc..

File: services/core-api/app/api/utils/access_decorators.py [hunk 5]
```diff
@@ -32,15 +35,15 @@
 
 
 def is_minespace_user():
-    return jwt.validate_roles([MINESPACE_PROPONENT])
+    return getJwtManager().validate_roles([MINESPACE_PROPONENT])
```
Comment (vyasworks) [inline services/core-api/app/api/utils/access_decorators.py:38]: going forward all JWT use should go through `getJwtManager()` which will pick the right jwt object based on the token in header.

File: services/core-api/app/extensions.py [hunk 6]
```diff
@@ -1,14 +1,54 @@
...
+
+# Gold SSO
+jwtv2 = JwtManager(None, os.environ.get('JWT_OIDC_WELL_KNOWN_CONFIG'), None, 'RS256', None, None, os.environ.get('JWT_OIDC_AUDIENCE'), None, None, False, False, None, JWT_ROLE_CALLBACK, None)
+# Existing Keycloak for integration clients
+jwtv1 = JwtManager(None, os.environ.get('JWT_OIDC_WELL_KNOWN_CONFIG_V1'), None, 'RS256', None, None, os.environ.get('JWT_OIDC_AUDIENCE_V1'), None, None, False, False, None, JWT_ROLE_CALLBACK_V1, None)
+# Test JWT Config for integration tests
+test_config = TestConfig()
+jwt = JwtManager(None, test_config.JWT_OIDC_WELL_KNOWN_CONFIG, None, 'RS256', None, None, test_config.JWT_OIDC_TEST_AUDIENCE, None, None, False, True, test_config.JWT_OIDC_TEST_KEYS, JWT_ROLE_CALLBACK, test_config.JWT_OIDC_TEST_PRIVATE_KEY_PEM)
+
+def getJwtManager():
```
Comment (vyasworks) [inline services/core-api/app/extensions.py:30]: picks jwt obj from above based on the API request, applies only to request context. 

objects remain in the app context.

File: services/core-api/app/extensions.py [hunk 7]
```diff
@@ -1,14 +1,54 @@
...
+    gold_sso = 'loginproxy.gov.bc.ca'
+    kc_realms = 'oidc.gov.bc.ca'
+    
+    auth_header = jwt.get_token_auth_header()
+    token = jwt_jose.get_unverified_claims(auth_header)
+
+    iss = token.get('iss')
+
+    if iss in test_config.JWT_OIDC_TEST_ISSUER:
```
Comment (vyasworks) [inline services/core-api/app/extensions.py:39]: spl config for testing.

File: services/core-api/app/extensions.py [hunk 8]
```diff
@@ -1,14 +1,54 @@
...
+def JWT_ROLE_CALLBACK(jwt_dict):
+    return (jwt_dict.get('client_roles') or [])
+
+def JWT_ROLE_CALLBACK_V1(jwt_dict):
+        return (jwt_dict['realm_access']['roles'])
+
 db = SQLAlchemy()
-jwt = JwtManager()
+
+# Gold SSO
```
Comment (vyasworks) [inline services/core-api/app/extensions.py:22]: New JWT objs go here as we get new integration clients - BCMI, GenTx etc..

File: services/core-api/app/flask_jwt_oidc_local/jwt_manager.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+from cachelib import SimpleCache
+from flask import _request_ctx_stack, current_app, g, jsonify, request
+from jose import jwt
+from six.moves.urllib.request import urlopen
+
+from .exceptions import AuthError
+
+
+class JwtManager:  # pylint: disable=too-many-instance-attributes
+    """Manages the JWT verification and JWKS key lookup."""
```
Comment (vyasworks) [inline services/core-api/app/flask_jwt_oidc_local/jwt_manager.py:33]: This is the main class that does jwt validation. 

forked and updated to support multi IDP. 

we could push this back to its own lib and remove it from our code base, but it was easier to develop here this way. 

Initially started with that approach but turned out to be complex with multiple pushes and iterations. 
I have a fork here https://github.com/bcgov/flask-jwt-oidc
and published here https://pypi.org/project/flask-jwt-oidc-mds/

this is a good potential improvement for future.

File: services/core-web/common/customAxios.js [hunk 10]
```diff
@@ -26,7 +26,7 @@ const CustomAxios = ({ errorToastMessage, suppressErrorNotification = false } =
 
       const status = error.response ? error.response.status : null;
       if (status === UNAUTHORIZED || status === MAINTENANCE) {
-        window.location.reload(false);
+        // window.location.reload(false);
```
Comment (matbusby-fw) [inline services/core-web/common/customAxios.js]:
```suggestion
        window.location.reload(false);
```

Comment (matbusby-fw) [reply services/core-web/common/customAxios.js]: I think we can restore this (likely removed while testing is my guess)

File: services/minespace-web/common/components/PartyAppointmentTable.js [hunk 11]
```diff
@@ -204,6 +208,7 @@ const mapDispatchToProps = (dispatch) =>
     {
       updatePartyRelationship,
       fetchPartyRelationships,
+      // getUserInfo,
```
Comment (matbusby-fw) [inline services/minespace-web/common/components/PartyAppointmentTable.js:211]: I'm wondering why this was removed?  Is this handled with in the new `/me` endpoint instead?

File: services/minespace-web/common/customAxios.js [hunk 12]
```diff
@@ -26,7 +26,7 @@ const CustomAxios = ({ errorToastMessage, suppressErrorNotification = false } =
 
       const status = error.response ? error.response.status : null;
       if (status === UNAUTHORIZED || status === MAINTENANCE) {
-        window.location.reload(false);
+        // window.location.reload(false);
```
Comment (matbusby-fw) [inline services/minespace-web/common/customAxios.js]:
```suggestion
        window.location.reload(false);
```

File: services/core-web/common/components/PartyAppointmentTable.js [hunk 13]
```diff
@@ -204,6 +208,7 @@ const mapDispatchToProps = (dispatch) =>
     {
       updatePartyRelationship,
       fetchPartyRelationships,
+      // getUserInfo,
```
Comment (taraepp) [inline services/core-web/common/components/PartyAppointmentTable.js:211]: need to verify this

File: services/minespace-web/src/utils/customAxios.js [hunk 14]
```diff
@@ -17,7 +17,7 @@ const CustomAxios = (errorToastMessage) => {
     (error) => {
       const status = error.response ? error.response.status : null;
       if (status === UNAUTHORIZED || status === MAINTENANCE) {
-        window.location.reload(false);
+        // window.location.reload(false);
```
Comment (matbusby-fw) [inline services/minespace-web/src/utils/customAxios.js:20]:
```suggestion
        window.location.reload(false);
```

File: services/core-web/common/customAxios.js (hunk 10 above)
Comment (vyasworks) [reply services/core-web/common/customAxios.js]: restored.

File: services/core-web/common/components/PartyAppointmentTable.js (hunk 13 above)
Comment (taraepp) [reply services/core-web/common/components/PartyAppointmentTable.js:211]: this can go. 👍

File: services/minespace-web/package.json [hunk 15]
```diff
@@ -182,5 +185,37 @@
   },
   "browserslist": [
     "> 0.1%"
-  ]
+  ],
+  "husky": {
```
Comment (taraepp) [inline services/minespace-web/package.json]: shouldn't be necessary, new husky version runs hooks from .husky folder in repo root

**Review by matbusby-fw — DISMISSED** [review]
BAM!

**Review by matbusby-fw — APPROVED** [review]

**Review by taraepp — APPROVED** [review]
👍

**Pull request merged**